lib/Sema/ConstraintSystem.h

```diff
@@ -3288,28 +3288,72 @@ class ConstraintSystem {
     /// The kind of bindings permitted.
     AllowedBindingKind Kind;
 
-    /// The kind of the constraint this binding came from.
-    ConstraintKind BindingSource;
-
-    /// The defaulted protocol associated with this binding.
-    ProtocolDecl *DefaultedProtocol;
-
-    /// If this is a binding that comes from a \c Defaultable constraint,
-    /// the locator of that constraint.
-    ConstraintLocator *DefaultableBinding = nullptr;
+  protected:
+    /// The source of the type information.
+    ///
+    /// Determines whether this binding represents a "hole" in
+    /// constraint system. Such bindings have no originating constraint
+    /// because they are synthetic, they have a locator instead.
+    PointerUnion<Constraint *, ConstraintLocator *> BindingSource;
 
     PotentialBinding(Type type, AllowedBindingKind kind,
-                     ConstraintKind bindingSource,
-                     ProtocolDecl *defaultedProtocol = nullptr,
-                     ConstraintLocator *defaultableBinding = nullptr)
+                     PointerUnion<Constraint *, ConstraintLocator *> source)
+        : BindingType(type->getWithoutParens()), Kind(kind),
+          BindingSource(source) {}
+
+  public:
+    PotentialBinding(Type type, AllowedBindingKind kind, Constraint *source)
         : BindingType(type->getWithoutParens()), Kind(kind),
-          BindingSource(bindingSource), DefaultedProtocol(defaultedProtocol),
-          DefaultableBinding(defaultableBinding) {}
+          BindingSource(source) {}
+
+    bool isDefaultableBinding() const {
+      if (auto *constraint = BindingSource.dyn_cast<Constraint *>())
+        return constraint->getKind() == ConstraintKind::Defaultable;
+      // If binding source is not constraint - it's a hole, which is
+      // a last resort default binding for a type variable.
+      return true;
+    }
+
+    bool hasDefaultedLiteralProtocol() const {
+      return bool(getDefaultedLiteralProtocol());
+    }
 
-    bool isDefaultableBinding() const { return DefaultableBinding != nullptr; }
+    ProtocolDecl *getDefaultedLiteralProtocol() const {
+      auto *constraint = BindingSource.dyn_cast<Constraint *>();
+      if (!constraint)
+        return nullptr;
+
+      return constraint->getKind() == ConstraintKind::LiteralConformsTo
+                 ? constraint->getProtocol()
+                 : nullptr;
+    }
+
+    ConstraintKind getSourceKind() const {
+      if (auto *constraint = BindingSource.dyn_cast<Constraint *>())
+        return constraint->getKind();
+      // If binding source is not constraint - it's a hole which is
+      // always a `Bind` constraint.
+      return ConstraintKind::Bind;
+    }
+
+    ConstraintLocator *getLocator() const {
```

Comment:
Just a question is that since we have a `getSourceKind()`, would this expected to be a `getSourceLocator()`?

Reply:
I think it's okay to name this locator because it really belongs to the binding, we need `getSourceKind()` to distinguish between kind of the binding binding accepts (represented by `Kind)` and kind of its source, which I'm not 100% certain is even necessary but I didn't want to extend the scope of changes here.

Reply:
It looks like we only use `getSourceKind` in one place and since we've got a locator, could probably use that instead there... Let me give it a try.

Reply:
`getSourceKind` has been removed :)